server/src/main/java/org/opensearch/index/recovery/RemoteStoreRestoreService.java

```diff
@@ -51,78 +57,86 @@ public class RemoteStoreRestoreService {
+        this.createIndexService = createIndexService;
+        this.metadataIndexUpgradeService = metadataIndexUpgradeService;
+        this.shardLimitValidator = shardLimitValidator;
     }
 
+    /**
+     * Restores data from remote store for indices specified in the restore request.
+     *
+     * @param request  restore request
+     * @param listener restore listener
+     */
     public void restore(RestoreRemoteStoreRequest request, final ActionListener<RestoreService.RestoreCompletionResponse> listener) {
         clusterService.submitStateUpdateTask("restore[remote_store]", new ClusterStateUpdateTask() {
             final String restoreUUID = UUIDs.randomBase64UUID();
             RestoreInfo restoreInfo = null;
 
-            @Override
-            public ClusterState execute(ClusterState currentState) {
-                // Updating cluster state
+            private ClusterState executeRestore(
+                ClusterState currentState,
+                Map<String, Tuple<Boolean, IndexMetadata>> indexMetadataMap,
+                boolean restoreAllShards
+            ) {
+                List<String> indicesToBeRestored = new ArrayList<>();
+                int totalShards = 0;
                 ClusterState.Builder builder = ClusterState.builder(currentState);
                 Metadata.Builder mdBuilder = Metadata.builder(currentState.metadata());
                 ClusterBlocks.Builder blocks = ClusterBlocks.builder().blocks(currentState.blocks());
                 RoutingTable.Builder rtBuilder = RoutingTable.builder(currentState.routingTable());
-
-                List<String> indicesToBeRestored = new ArrayList<>();
-                int totalShards = 0;
-                for (String index : request.indices()) {
-                    IndexMetadata currentIndexMetadata = currentState.metadata().index(index);
-                    if (currentIndexMetadata == null) {
-                        // ToDo: Handle index metadata does not exist case. (GitHub #3457)
-                        logger.warn("Remote store restore is not supported for non-existent index. Skipping: {}", index);
-                        continue;
+                for (Map.Entry<String, Tuple<Boolean, IndexMetadata>> indexMetadataEntry : indexMetadataMap.entrySet()) {
+                    String indexName = indexMetadataEntry.getKey();
+                    IndexMetadata indexMetadata = indexMetadataEntry.getValue().v2();
+                    boolean fromRemoteStore = indexMetadataEntry.getValue().v1();
+                    IndexMetadata updatedIndexMetadata = indexMetadata;
+                    Map<ShardId, ShardRouting> activeInitializingShards = new HashMap<>();
+                    if (restoreAllShards) {
+                        updatedIndexMetadata = IndexMetadata.builder(indexMetadata)
+                            .state(IndexMetadata.State.OPEN)
+                            // do we need to increment this during restore from remote index metadata
```

Comment:
Also assert in case the index exists locally as well, it is closed and its version would be <= than remote index metadata versions.

Reply:
> lets increment so that it clearly indicates this version increment was restore from remote
Sure, will update to increment for remote cluster state restore as well